### PR TITLE
Ensure that we always mark an analyzer file reference full path as a

### LINKED ISSUE #32226: Analyzer loader fails to load dll that lives next to analyzer
**Version Used**: 

**Steps to Reproduce**:

1. git clone http://github.com/dotnet/project-system
2. cd project-system
3. git checkout eacf50af576d7016b40e5b880c18cc93ee22f258
4. build
5. ProjectSystem.sln

**Expected Behavior**:
No errors about loading binaries

**Actual Behavior**:

```
Severity	Code	Description	Project	File	Line	Category	Suppression State	Detail Description
Warning	CS8032	An instance of analyzer Roslyn.Diagnostics.CSharp.Analyzers.CSharpCodeActionCreateAnalyzer cannot be created from C:\Users\davkean\.nuget\packages\roslyn.diagnostics.analyzers\2.6.3-beta1.18614.2\analyzers\dotnet\cs\Roslyn.Diagnostics.CSharp.Analyzers.dll: Could not load file or assembly 'Roslyn.Diagnostics.Analyzers, Version=2.6.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..	Microsoft.VisualStudio.ProjectSystem.Managed.VS		1	Roslyn.HostError	Active	Could not load file or assembly 'Roslyn.Diagnostics.Analyzers, Version=2.6.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
-----
Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework\v4.0.30319\clr.dll
Running under executable  C:\Program Files (x86)\Microsoft Visual Studio\2019\16.0stg\Common7\IDE\devenv.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = Roslyn.Diagnostics.Analyzers, Version=2.6.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
 (Fully-specified)
LOG: Appbase = file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/
LOG: Initial PrivatePath = NULL
Calling assembly : Roslyn.Diagnostics.CSharp.Analyzers, Version=2.6.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: C:\Users\davkean\AppData\Local\Microsoft\VisualStudio\16.0_afa0bdd1\devenv.exe.config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework\v4.0.30319\config\machine.config.
LOG: Post-policy reference: Roslyn.Diagnostics.Analyzers, Version=2.6.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PublicAssemblies/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PublicAssemblies/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TemplateProviders/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TemplateProviders/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TestWindow/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TestWindow/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/XamlLanguageService/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/XamlLanguageService/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/Debugger/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/Debugger/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/DiagnosticsHub/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/DiagnosticsHub/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/x86/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/x86/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PublicAssemblies/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PublicAssemblies/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TemplateProviders/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TemplateProviders/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TestWindow/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/TestWindow/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/XamlLanguageService/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Microsoft/XamlLanguageService/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/Debugger/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/Debugger/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/DiagnosticsHub/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/CommonExtensions/Platform/DiagnosticsHub/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/x86/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2019/16.0stg/common7/ide/PrivateAssemblies/DataCollectors/x86/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Users/davkean/AppData/Local/Temp/VS/AnalyzerAssemblyLoader/36ea2f77fde84a02b288afa6ce94ba22/12/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Users/davkean/AppData/Local/Temp/VS/AnalyzerAssemblyLoader/36ea2f77fde84a02b288afa6ce94ba22/12/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.DLL.
LOG: Attempting download of new URL file:///C:/Users/davkean/AppData/Local/Temp/VS/AnalyzerAssemblyLoader/36ea2f77fde84a02b288afa6ce94ba22/12/Roslyn.Diagnostics.Analyzers.EXE.
LOG: Attempting download of new URL file:///C:/Users/davkean/AppData/Local/Temp/VS/AnalyzerAssemblyLoader/36ea2f77fde84a02b288afa6ce94ba22/12/Roslyn.Diagnostics.Analyzers/Roslyn.Diagnostics.Analyzers.EXE.

-----


```

If I look at the "shadow caching" (I presume), I see:

```
> dir C:\Users\davkean\AppData\Local\Temp\VS\AnalyzerAssemblyLoader\36ea2f77fde84a02b288afa6ce94ba22\12

Roslyn.Diagnostics.CSharp.Analyzers.dll
```

If I look at the package, I see:

```
> dir C:\Users\davkean\.nuget\packages\roslyn.diagnostics.analyzers\2.6.3-beta1.18614.2\analyzers\dotnet\cs

Roslyn.Diagnostics.Analyzers.dll
Roslyn.Diagnostics.CSharp.Analyzers.dll
```

### PR DESCRIPTION
**Pull request by mavasani**
dependency location, otherwise we might fail to load the assembly if any of its _referencing_ analyzer assemblies are loaded first.

Fixes #32226

### CODE REVIEW

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            var analyzerDependencyFile = CreateAnalyzerDependency();
+            var analyzerMainFile = CreateMainAnalyzerWithDependency(analyzerDependencyFile);
+            var loader = new ShadowCopyAnalyzerAssemblyLoader(Path.Combine(directory.Path, "AnalyzerAssemblyLoader"));
+
+            var analyzerMainReference = new AnalyzerFileReference(analyzerMainFile.Path, loader);
+            analyzerMainReference.AnalyzerLoadFailed += (_, e) => AssertEx.Fail(e.Exception.Message);
+            var analyzerDependencyReference = new AnalyzerFileReference(analyzerDependencyFile.Path, loader);
+            analyzerDependencyReference.AnalyzerLoadFailed += (_, e) => AssertEx.Fail(e.Exception.Message);
+
+            var analyzers = analyzerMainReference.GetAnalyzersForAllLanguages();
```
Comment (mavasani) [inline src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: Verified that the test fails prior to the product code change with `Message: Could not load file or assembly 'AnalyzerDependency, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified`

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 2]
```diff
@@ -46,19 +46,13 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
...
-
-            _fullPath = fullPath;
+            _fullPath = fullPath ?? throw new ArgumentNullException(nameof(fullPath));
             _diagnosticAnalyzers = new Extensions<DiagnosticAnalyzer>(this, IsDiagnosticAnalyzerAttribute);
-            _assemblyLoader = assemblyLoader;
+            _assemblyLoader = assemblyLoader ?? throw new ArgumentNullException(nameof(assemblyLoader));
+
+            // Note this analyzer full path as a dependency location, so that the analyzer loader
+            // can correctly load analyzer dependencies.
+            assemblyLoader.AddDependencyLocation(fullPath);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs]: This is the only _real_ change that fixes the bug.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 3]
```diff
@@ -34,12 +37,12 @@ public void AddDependencyLocation(string fullPath)
             {
                 if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths))
                 {
-                    _knownAssemblyPathsBySimpleName.Add(simpleName, new List<string>() { fullPath });
-                }
-                else if (!paths.Contains(fullPath))
-                {
-                    paths.Add(fullPath);
+                    // Use case sensitive string comparer for file paths.
+                    paths = new HashSet<string>(StringComparer.Ordinal);
```
Comment (genlu) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Do we need to use `Roslyn.Utilities.PathUtilities.Comparer`?

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Done, thanks.

**Comment by genlu** [comment]
Could you please explain to me how analyzer loading works, I don't understand what has caused this problem. Thanks! :)

**Comment by mavasani** [comment]
Ping @dotnet/roslyn-compiler for reviews.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 4]
```diff
@@ -46,19 +46,16 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
...
-                throw new ArgumentNullException(nameof(fullPath));
-            }
+            _fullPath = fullPath ?? throw new ArgumentNullException(nameof(fullPath));
+            _diagnosticAnalyzers = new Extensions<DiagnosticAnalyzer>(this, IsDiagnosticAnalyzerAttribute);
+            _assemblyLoader = assemblyLoader ?? throw new ArgumentNullException(nameof(assemblyLoader));
 
-            if (assemblyLoader == null)
+            // Note this analyzer full path as a dependency location, so that the analyzer loader
+            // can correctly load analyzer dependencies.
+            if (PathUtilities.IsAbsolute(fullPath))
```
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs]: >PathUtilities.IsAbsolute(fullPath)) [](start = 16, length = 35)

Is this check necessary?

Comment (cston) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs]: Thanks. Option 2, the current approach, seems reasonable.

**Comment by mavasani** [comment]
Ping @dotnet/roslyn-compiler for an additional review.

@jinujoseph @vatsalyaagrawal for approval

**Comment by jinujoseph** [comment]
@jcouv  @333fred  need second compiler review

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using Xunit;
+
+// The ShadowCopyAnalyzerAssemblyLoader type is only defined for the below platforms
+#if NET472 || NETCOREAPP1_1 || NETCOREAPP2_1
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: `NETCOREAPP1_1` is not used anymore in testing. It shouldn't be needed here.

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                   },
+                   new CSharp.CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+                return directory.CreateDirectory("AnalyzerMain").CreateFile("AnalyzerMain.dll").WriteAllBytes(analyzerMainCompilation.EmitToArray());
+            }
+        }
+    }
+}
+
+#endif
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: Consider adding the following:

```
#else
#error unsupported configuration
#endif
```

This helps guard against future changes to the target frameworks that roslyn uses. Consider that in the near future netcoreapp3.0 will be a valid target framework. With this proposed change we would get a compiler error if we forgot to update this file. Without the test would just silently stop running. 

I've been trying to take this approach in all places where we have `#if` as I've gotten this wrong several times.

File: src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs [hunk 7]
```diff
@@ -7,6 +7,9 @@
 using System.Threading;
 using System.Threading.Tasks;
 
+// The ShadowCopyAnalyzerAssemblyLoader derives from DesktopAnalyzerAssemblyLoader (NET472) OR CoreClrAnalyzerAssemblyLoader (NETCOREAPP1_1 || NETCOREAPP2_1)
+#if NET472 || NETCOREAPP1_1 || NETCOREAPP2_1
```
Comment (jaredpar) [inline src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: We don't compile for NETCOREAPP1_1 anymore so this shouldn't be needed.

File: src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs [hunk 8]
```diff
@@ -195,3 +198,5 @@ private string CreateUniqueDirectoryForAssembly()
         }
     }
 }
+
+#endif
```
Comment (jaredpar) [inline src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: Please add the same `#else` pattern I recomended in the other file to proof against future changes.

File: src/Test/Utilities/Portable/Roslyn.Test.Utilities.csproj [hunk 9]
```diff
@@ -77,6 +77,9 @@
     <Compile Include="..\..\..\Compilers\Shared\DesktopAnalyzerAssemblyLoader.cs">
       <Link>DesktopAnalyzerAssemblyLoader.cs</Link>
     </Compile>
+    <Compile Include="..\..\..\Compilers\Shared\ShadowCopyAnalyzerAssemblyLoader.cs">
+      <Link>ShadowCopyAnalyzerAssemblyLoader.cs</Link>
```
Comment (jaredpar) [inline src/Test/Utilities/Portable/Roslyn.Test.Utilities.csproj]: Nit: you can avoid the Link element here if you like. It's not actually needed unelss you want to change the file name. We should probably just clean all this up.

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs (hunk 6 above)
Comment (mavasani) [reply src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: Thanks, done.

**Comment by mavasani** [comment]
@jaredpar Address feedback, thanks!

**Pull request merged**